models/imap.go

```diff
@@ -0,0 +1,177 @@
+package models
+
+import (
+	"errors"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/glennzw/eazye"
+	log "github.com/gophish/gophish/logger"
+)
+
+// IMAP contains the attributes needed to handle logging into an IMAP server to check
+// for reported emails
+type IMAP struct {
+	UserId            int64     `json:"-" gorm:"column:user_id"`
+	Enabled           bool      `json:"enabled"`
+	Host              string    `json:"host"`
+	Port              uint16    `json:"port,string,omitempty"`
+	Username          string    `json:"username"`
+	Password          string    `json:"password"`
+	TLS               bool      `json:"tls"`
+	Folder            string    `json:"folder"`
+	RestrictDomain    string    `json:"restrict_domain"`
+	DeleteCampaign    bool      `json:"delete_campaign"`
+	LastLogin         time.Time `json:"last_login,omitempty"`
+	LastLoginFriendly string    `json:"last_login_friendly,omitonempty"`
+	ModifiedDate      time.Time `json:"modified_date"`
+}
+
+// ErrIMAPHostNotSpecified is thrown when there is no Host specified
+// in the IMAP configuration
+var ErrIMAPHostNotSpecified = errors.New("No IMAP Host specified")
+
+// ErrIMAPPortNotSpecified is thrown when there is no Port specified
+// in the IMAP configuration
+var ErrIMAPPortNotSpecified = errors.New("No IMAP Port specified")
+
+// ErrInvalidIMAPHost indicates that the IMAP server string is invalid
+var ErrInvalidIMAPHost = errors.New("Invalid IMAP server address")
+
+// ErrInvalidIMAPPort indicates that the IMAP Port is invalid
+var ErrInvalidIMAPPort = errors.New("Invalid IMAP Port")
+
+// ErrIMAPUsernameNotSpecified is thrown when there is no Username specified
+// in the IMAP configuration
+var ErrIMAPUsernameNotSpecified = errors.New("No Username specified")
+
+// ErrIMAPPasswordNotSpecified is thrown when there is no Password specified
+// in the IMAP configuration
+var ErrIMAPPasswordNotSpecified = errors.New("No Password specified")
+
+// TableName specifies the database tablename for Gorm to use
+func (s IMAP) TableName() string {
+	return "imap"
+}
+
+// Validate ensures that IMAP configs/connections are valid
+func (s *IMAP) Validate() error {
+	switch {
+	case s.Host == "":
+		return ErrIMAPHostNotSpecified
+	case s.Port == 0:
+		return ErrIMAPPortNotSpecified
+	case s.Username == "":
+		return ErrIMAPUsernameNotSpecified
+	case s.Password == "":
+		return ErrIMAPPasswordNotSpecified
+	}
+
+	// Set the default value for Folder
+	if s.Folder == "" {
+		s.Folder = "INBOX"
```

Comment:
I'd recommend moving this to a constant. Something like:

```go
const DefaultIMAPFolder = "INBOX"
```

Reply:
Done